### ISSUE TITLE
C output: Change from Doubles to Floats

### ISSUE THREAD

**Issue by travisfranck** [climateinteractive/SDEverywhere#7]
We use 64-bit math in SDEverywhere because of a long-standing rule in C that all internal math operations are 64 bit, even if the operands are floats, and results are ultimately rounded to 32 bit precision. I elected to maintain 64-bit precision throughout by storing data in C doubles.

But your comment led me to look into this further yesterday. I discovered that in the C99 standard that modern compilers like clang use, the precision of internal math operations is implementation dependent. And guess what? Clang does not enforce 64 bit precision internally. If you use float operands, math operations are 32-bit.

I think moving to 32-bit math makes sense for SDEverywhere. That is enough precision for SD work. Most people use the 32 bit version of Vensim. We could remove the requirement for validating against data from 64-bit Vensim. And we would cut our memory requirements for data in half. This mostly involves changing "double" to "float" everywhere and using the float versions of standard library functions.

-Todd

**Comment by ToddFincannon** [comment]
It also turns out that Vensim stores all values as floats, even when it does double-precision computations. https://www.vensim.com/documentation/index.html?ref_variable_names.htm

**Comment by chrispcampbell** [comment]
I did a quick experiment to change all uses of doubles to floats (actually an `sde_float` typedef) in SDEverywhere c files (and in the generated c file).

There are some good performance and size wins to be had from this change (see below), but we'd need to spend some time making sure that any differences in the new outputs are within an acceptable range, so I think we should hold off on the change until we get some more automated checks in place.  The runtime performance gains appear to be very slight on modern processors; the main benefit appears to be in shaving off ~120kb from the wasm binary size.

Here are some performance numbers for the record relative to other performance work.

### Performance

__MacBook Pro (2019)__ | _2.4 GHz 8-core i9, 32 GB RAM, macOS 10.15_

| Issue | C run (ms) | Wasm run (ms) | Wasm init (ms) | JS mem (MB) | Page mem (MB) |
| ----- | ---------: | ------------: | -------------: | ----------: | ------------: |
| baseline | 45.8 | 87.5 | 38.0 | 94 | 685 |
| SDE 18   | 46.0 | 85.6 | 18.0 | 39 | 672 |
| SDE 19   | 42.8 | 49.4 | 15.0 | 38 |  25 |
| SDE 22   | 34.8 | 44.8 | 15.0 | 38 |  21 |
| SDE 23   | 32.7 | 42.8 | 13.0 | 38 |  32 |
| SDE 24   | 26.6 | 38.2 | 13.0 | 39 |  26 |
| SDE 7    | 24.4 | 34.8 | 10.0 | 39 |  20 |

__iPhone 8__ | _A11, iOS 13_

| Issue | C run (ms) | Wasm run (ms) | Wasm init (ms) | JS mem (MB) | Page mem (MB) |
| ----- | ---------: | ------------: | -------------: | ----------: | ------------: |
| baseline | 39.9 | 187.0 | 165.0 | 39 | 645 |
| SDE 18   | 40.3 | 219.0 |  86.0 | 38 | 724 |
| SDE 19   | 40.1 |  81.6 |  83.0 | 38 |  41 |
| SDE 22   | 35.5 |  74.6 |  86.0 | 40 |  40 |
| SDE 23   | 31.1 |  73.6 |  82.0 | 41 |  39 |
| SDE 24   | 28.5 |  71.6 |  82.0 | 40 |  38 |
| SDE 7    | 28.7 |  70.0 |  70.0 | 36 |  36 |

__iPad Air (2013)__ | _A7, iOS 12_

| Issue | C run (ms) | Wasm run (ms) | Wasm init (ms) | JS mem (MB) | Page mem (MB) |
| ----- | ---------: | ------------: | -------------: | ----------: | ------------: |
| baseline | 151.0 | 1372.2 | 30146.0 | 77 | 331 |
| SDE 18   | 166.0 | 1408.0 |  4416.0 | 42 | 395 |
| SDE 19   | 151.0 |  837.6 |  1291.0 | 45 |  41 |
| SDE 22   | 137.0 |  771.6 |  1484.0 | 44 |  40 |
| SDE 23   | 110.1 |  642.2 |  1148.0 | 44 |  41 |
| SDE 24   | 111.8 |  638.4 |  1236.0 | 45 |  37 |
| SDE 7    |  91.7 |  543.8 |  1120.0 | 43 |  51 |

### Size

| Issue | Wasm size (bytes) |
| ----- | ----------------: |
| baseline | 1,084,036 |
| SDE 18   |   773,968 |
| SDE 19   |   776,851 |
| SDE 22   |   737,028 |
| SDE 23   |   741,668 |
| SDE 24   |   741,677 |
| SDE 7    |   616,264 |

### Legend

| Issue | Date | Notes |
| ----- | ---- | ----- |
| baseline | 2020/07/08 | baseline prior to performance work |
| [SDE 18](https://github.com/climateinteractive/SDEverywhere/issues/18) | 2020/07/09 | change lookup init to use static arrays |
| [SDE 19](https://github.com/climateinteractive/SDEverywhere/issues/19) | 2020/07/09 | break large functions into chunked subfunctions |
| [SDE 22](https://github.com/climateinteractive/SDEverywhere/issues/22) | 2020/07/10 | replace wrapper functions with macros |
| [SDE 23](https://github.com/climateinteractive/SDEverywhere/issues/23) | 2020/07/10 | replace dimension array access with simple index |
| [SDE 24](https://github.com/climateinteractive/SDEverywhere/issues/24) | 2020/07/10 | optimize __lookup function |
| [SDE 7](https://github.com/climateinteractive/SDEverywhere/issues/7)   | 2020/07/10 | change from doubles to floats |

**Comment by chrispcampbell** [comment]
I've experimented with floats vs doubles a few times since my last comment in 2020.  Every time I try (with En-ROADS and/or C-ROADS), I would see a modest speed improvement but at the cost of precision.

The latter _might_ be acceptable in many cases, but if we switched to using floats everywhere, there is a risk of getting into a situation where there is a significant loss of precision that really affects how outputs are interpreted and we wouldn't know (unless we build both double and float variants of the model and compared the two often).

I had Claude do a fresh analysis today using En-ROADS as the test case.

TL;DR: En-ROADS (and likely other models) could see a ~20% improvement in model run time, but it comes at the cost of reduced precision which affects some outputs more than others.

I think that given these trade-offs, it's not worth pursuing this change in SDE.  We could add C preprocessor flags/macros to make it possible for users to switch at build time if they are willing to accept the trade-offs, but that would be a fair amount of work with not much to gain.  Therefore, I think we should close this as "investigated but will not fix".

I've attached Claude's notes below.  I've also attached the experiment scripts/sources that Claude created in a zip file (see [floats.zip](https://github.com/user-attachments/files/31243356/floats.zip)).

<details><summary>Click to show Claude's notes</summary>
<p>

# SDEverywhere #7 — doubles vs floats, measured on En-ROADS

Experiment run 2026-08-19 against `en-roads-app1/model/app-prep/sde-prep/build`
(455 output variables, 300 inputs, INITIAL TIME 1990, FINAL TIME 2100,
TIME STEP 0.125, SAVEPER 1 → 880 steps, 111 save points).

## Method

Two copies of the generated C sources (`processed.c`, `model.c`, `vensim.c`,
`macros.c` + headers):

- **dbl** — unmodified.
- **flt** — mechanically converted:
  1. `double` → `float` everywhere (including `_time` and the control params);
  2. `DBL_MAX`/`DBL_MIN` → `FLT_MAX`/`FLT_MIN` (the `_NA_` sentinel);
  3. `<tgmath.h>` included from `sde.h` so `pow`/`exp`/`log`/`fmod`/… dispatch to
     the `f` variants;
  4. every floating-point literal given an `f` suffix.

Step 4 matters. Without it, a `float * 2.0` expression promotes to double and
clang emits calls to `pow`/`exp`/`log` (verified with `nm -u`), which costs about
a third of the speedup. A real code-generator change would emit `f` literals, so
step 4 is the faithful configuration. Both variants are reported below.

`TIME STEP` is 0.125, a power of two, so `_time` accumulates **exactly** in
float32 across the whole 1990–2100 range. The `fmod(_time, _saveper) < 1e-6`
save-point test and the `_time == _final_time` test behave identically in both
builds — no save points are lost or gained. Time bookkeeping is a non-issue for
En-ROADS specifically; it would not be for a model with e.g. TIME STEP = 0.1.

Benchmarks: 30 timed iterations after a warm-up, `runModelWithBuffers` only
(no I/O), median of three separate process runs. Native builds use
`-ffp-contract=off` so that native and Wasm do the same arithmetic (Wasm has no
FMA). Machine: Apple Silicon, macOS 15.6, Apple clang 16, emcc 3.1.46, Node 22
(V8). Wasm is built with the app's real `emcc-args.txt` and driven from JS the
same way `packages/runtime` does it (`_malloc` + HEAP views + `cwrap`).

## Speed

| Configuration                                      |      double |       float |   change |
| -------------------------------------------------- | ----------: | ----------: | -------: |
| Native, clang `-O2`                                |     13.7 ms | **10.4 ms** | **−24%** |
| Native, clang `-O2`, float literals _not_ suffixed |     13.7 ms |     11.1 ms |     −19% |
| **Wasm, `-O2` (current app config)**               | **29.4 ms** | **23.1 ms** | **−21%** |
| Wasm, `-O3 -msimd128`                              |     23.1 ms |     19.3 ms |     −17% |

Wasm module init was 5.1–5.9 ms (double) vs 4.6–5.3 ms (float) — no meaningful
change.

Note that `-msimd128` does **not** favour floats disproportionately (4-wide f32
vs 2-wide f64 buys nothing here). En-ROADS is dominated by scalar dependency
chains, not vectorizable loops, so the float win is essentially the same at every
optimization level. The float gain and the SIMD gain are largely additive:
`-O2` double 29.4 ms → `-O3 -msimd128` float 19.3 ms is −34% overall.

## Size

| Artifact                           |    double |     float | change |
| ---------------------------------- | --------: | --------: | -----: |
| `.wasm`, `-O2`                     |   866,671 |   759,707 | −12.3% |
| `.wasm`, `-O2`, gzipped            |   301,065 |   276,828 |  −8.1% |
| Shipped `SINGLE_FILE` `.js` bundle | 1,167,309 | 1,024,689 | −12.2% |
| Shipped bundle, gzipped            |   391,266 |   361,543 |  −7.6% |
| `.wasm`, `-O3 -msimd128`           | 1,099,219 |   880,451 | −19.9% |

Broken down, the `-O2` `.wasm`:

| Section                         |  double |   float | change |
| ------------------------------- | ------: | ------: | -----: |
| code                            | 737,949 | 697,164 |  −5.5% |
| data (lookup tables, constants) | 128,034 |  61,861 | −51.7% |

Native object sections tell the same story for runtime state: `__data`
116,020 → 58,012 and `__common` (model variable storage) 209,512 → 105,976 —
both essentially halved. On the JS side the output buffer would become a
`Float32Array`, halving that too (455 × 111 × 8 = 404 KB → 202 KB per run).

Historical note: the numbers in issue #7 from 2020/07 (1,084,036 → 616,264 bytes)
are not comparable — that was a much smaller model, an older Emscripten, and
`-Os`. The proportional size win today is ~12%, not ~43%.

## Accuracy

Two error metrics per value:

- **relOwn** = |Δ| / |baseline| — pointwise relative error.
- **relRange** = |Δ| / (max |baseline| over that variable's series) — how visible
  the error would be if the series were plotted. This is the metric that matters
  for the app and for model-check thresholds.

`:NA:` sentinels and the ~20 values the model itself produces as NaN are excluded
(they are identical in both builds unless noted).

### Well-behaved scenarios

Default (all sliders at defaults), all-min, all-max, all-sliders-at-3/4, and the
seven NGFS policy scenarios:

| Scenario                                                         | vars > 1% of range | > 0.1% | median relOwn | p99 relOwn |
| ---------------------------------------------------------------- | -----------------: | -----: | ------------: | ---------: |
| default                                                          |            0 / 454 |      5 |        1.5e-6 |     4.6e-4 |
| all-min                                                          |            0 / 454 |      4 |             — |          — |
| all-max                                                          |            0 / 455 |      0 |             — |          — |
| three-quarter                                                    |            0 / 455 |      0 |             — |          — |
| ngfs-cp / ndc / fragmented / delayed / 2c / low-demand / nz-2050 |            0 / 454 |      2 |             — |          — |

In these scenarios the worst chart-visible error is ~0.15–0.3% and always on a
near-zero series (e.g. `Final hydrogen for energy carrier…` at 1e-6 magnitude).
Nothing here would be noticed in the UI.

### Pathological scenarios

`alt-a` (sliders alternating min/max) is a different story:

|                                        |        value |
| -------------------------------------- | -----------: |
| output vars with worst relRange > 1%   | **46 / 455** |
| output vars with worst relRange > 0.1% |    156 / 455 |
| max relRange                           |     **0.82** |

The mechanism, on the worst variable:

```
year   double     float
2034   0.00130   0.00133
2035   0.01684   0.00770
2036   0.40220   0.04172
2037  12.09582   0.36490     <- 82% of the series range
2038  14.21279   3.67359
2039  14.19078  12.84468
2040  14.17541  14.06926
2042  14.18012  14.17356     <- converged
        (Percent share of transport new capital that uses hydrogen)
```

This is a technology-adoption S-curve. Accumulated float rounding delays the
takeoff by roughly one to two years; both runs converge to the same ~14.2%
plateau. It is a **timing shift on a near-vertical transition**, not a divergent
answer — but on a chart for those years it is unmistakable.

A second mechanism shows up as `:NA:` sentinel flips. In `alt-a`,
`Hydrogen TCO vs weighted average of alternative equipment` is `:NA:` in the
double build for 2067–2076 but a real number (~0.29) in the float build — a
`XIDZ`/`ZIDZ`-style guard firing differently. The app would draw a gap in one
build and a line in the other. Ten such values in `alt-a`; 5–8 of 60 scenarios
in the sweep below have at least one.

### How often does this happen?

120 randomized scenarios (all 216 numeric sliders set to a random in-range,
step-snapped value; two seeds):

|                                                   |     seed 1234 |      seed 777 |
| ------------------------------------------------- | ------------: | ------------: |
| scenarios with ≥1 var differing > 1% of its range | 15 / 60 (25%) |  9 / 60 (15%) |
| scenarios with ≥1 var differing > 0.1%            | 54 / 60 (90%) | 45 / 60 (75%) |
| scenarios with ≥1 var differing > 0.01%           |       60 / 60 |       60 / 60 |
| worst single relRange                             |          0.22 |          0.45 |
| max vars > 1% in one scenario                     |            31 |            75 |
| scenarios with `:NA:` sentinel mismatches         |        8 / 60 |        5 / 60 |

So: **roughly one in five randomly-chosen scenarios has at least one output
variable that differs visibly (>1% of its own range)**, and the specific
variables involved are prominent ones — `Market price of oil`,
`Total direct use carriers for each end use`,
`Percent share of transport new capital that uses hydrogen`,
`Long term storage capacity`, `Damage`.

Headline climate outputs are far more robust. In a scenario where
`Market price of oil` differed by 9.5% of range:

| Variable                     | worst relRange |
| ---------------------------- | -------------: |
| Temperature change from 1850 |         5.3e-5 |
| CO2 Equivalent Net Emissions |         3.9e-3 |
| Market price of oil          |         9.5e-2 |

The oil-price case is worth understanding: in that scenario the _double_ run is
itself oscillating hard (2078: 170.4, 2079: 153.5, 2080: 147.3, 2081: 160.1,
2082: 174.0, 2083: 149.1). Float shifts the phase of an oscillation the model is
already producing. That regime is arguably a pre-existing model/TIME STEP issue
that floats merely expose.

### Reproducibility across toolchains

Same source, native clang vs Emscripten/V8:

|                        | max relRange (default) | max relRange (`alt-a`) |
| ---------------------- | ---------------------: | ---------------------: |
| double: native vs Wasm |                1.5e-13 |                8.4e-13 |
| float: native vs Wasm  |                 8.2e-5 |                 1.1e-3 |

This is the finding I'd weight most heavily. With doubles the two backends agree
to ~13 digits and reference data is portable. With floats, differences in
`powf`/`expf`/`logf` implementations and instruction selection make results
**toolchain-dependent at the 1e-4 level** — the same order as the double-vs-float
difference itself in benign scenarios. Any stored baseline, model-check
threshold, or Vensim comparison would have to tolerate that, and a compiler or
emsdk upgrade could shift results.

Reassuringly, the double-vs-float conclusions are identical whether measured
natively or in Wasm (Wasm `alt-a`: 46 vars > 1%, max relRange 0.82 — matching
native to 3 digits).

## Conditioning check

Perturbing the _inputs_ of `alt-a` by relative 1e-7 … 1e-3 in the double build
moves the hydrogen S-curve by only ~3.5% at its steepest point, versus ~90% for
float. Float error is not equivalent to a small input perturbation: it is
injected at every operation and compounds across 880 time steps and thousands of
intermediate variables. (This particular sweep is partly confounded — several
En-ROADS inputs are switches or year thresholds, so the perturbed runs all snap
to the same discrete branch — but the order-of-magnitude conclusion holds.)

## Bottom line

- **~21% faster in Wasm** (29.4 → 23.1 ms), ~24% native, ~12% smaller binary,
  model state memory halved. Real but not transformative.
- The 2020 recollection of "significant differences" is confirmed and now
  quantified: **~20% of random scenarios have at least one output variable that
  moves by more than 1% of its range, worst case 82%.** The mechanism is
  amplification at S-curve transitions and `:NA:` guard flips, not uniform
  precision loss — headline temperature/emissions outputs stay within 5e-5/4e-3.
- The under-appreciated cost is **loss of cross-toolchain reproducibility**
  (1e-4 native-vs-Wasm with floats, vs 1e-13 with doubles).
- If this is pursued, the useful intermediate is a per-model or per-variable
  opt-in rather than a global switch, plus a model-check comparison run over a
  scenario sweep like the one above — the 2020 conclusion ("hold off until we
  have automated checks in place") still reads correctly, and those checks now
  exist.

</p>
</details>
